File: test/tests/inheritance.js
const { assert } = require('chai');
const { mocha_object_store } = require('../opts');

it('nominal', () => {
  const b = new dll.Base(12);
  assert.instanceOf(b, dll.Base);
  assert.notInstanceOf(b, dll.Derived);
  assert.strictEqual(b.get(), 12);
  assert.strictEqual(b.base_get(), 12);

  const d = new dll.Derived(10);
  assert.instanceOf(d, dll.Derived);
  assert.instanceOf(d, dll.Base);
  assert.strictEqual(d.get(), 11);
  assert.strictEqual(d.virtual_base_get(), 11);
  assert.strictEqual(d.base_get(), 10);
  assert.strictEqual(d.derived_get(), 11);

  assert.throws(() => {
    // @ts-expect-error
    b.derived_get();
  }, /b\.derived_get is not a function/);

  assert.strictEqual(d.ret1(), 1);
  assert.strictEqual(d.ret2(), 2);
});

it('abstract', () => {
  const d = new dll.DerivedAbstract(17);
  assert.instanceOf(d, dll.DerivedAbstract);
  assert.instanceOf(d, dll.Abstract);
  assert.strictEqual(d.id(), 17);

  assert.throws(() => {
    // @ts-expect-error
    new dll.Abstract;
  }, /create an object of abstract/);
});

it('automatic upcasting', () => {
  const derived = new dll.Derived(10);

  // Derived is accepted as Base but works as Derived
  assert.strictEqual(dll.requireBase(derived), 11);
});

it('type distinction by the object store', () => {
  if (!mocha_object_store())
    return;
  const derived = new dll.Derived(12);

  // returnBase returns const Base & which creates a new wrapper
  // which is not of the Derived type
  const ret = dll.returnBase(derived);
  assert.instanceOf(ret, dll.Base);
  assert.notInstanceOf(ret, dll.Derived);

  // However the C++ object inside is Derived
  assert.strictEqual(ret.get(), 13);
});

it('object store object reuse', () => {
  if (!mocha_object_store())
    return;

  const base = new dll.Base(16);

  const ret = dll.returnBase(base);
  assert.instanceOf(ret, dll.Base);
  assert.notInstanceOf(ret, dll.Derived);

  // returnBase should not create a new wrapper
  assert.strictEqual(ret, base);

  assert.strictEqual(ret.get(), 16);
});

File: test/tests/inheritance.cc
#include <fixtures/abstract.h>
#include <fixtures/inheritance.h>

#include <nobind.h>

NOBIND_MODULE(inheritance, m) {
  m.def<IF1>("IF1").def<&IF1::ret1>("ret1");
  m.def<IF2>("IF2").def<&IF2::ret2>("ret2");
  m.def<Base>("Base").cons<int>().def<&Base::get>("get").def<&Base::base_get>("base_get");

  m.def<Derived, Base, IF1, IF2>("Derived")
      .cons<int>()
      .def<&Derived::get>("get")
  // Both of these are artificial edge cases for testing purposes
  // and MSVC 2019 is already pushed to its limits, we choose to certify
  // it without them as they are not really needed
#if !defined(_MSC_VER) || _MSC_VER >= 1930
      // Get explicitly the address to the virtual method in the Base class
      // The C++ templating will still invoke the correct method as the
      // invocation happens through a Derived object
      .def<&Base::get>("virtual_base_get")
      .def<&Derived::base_get>("base_get")
      .def<&IF1::ret1>("ret1")
      .def<&IF2::ret2>("ret2")
#else
      .def<&Derived::get>("virtual_base_get")
      .def<static_cast<int (Derived::*)() const>(&Derived::base_get)>("base_get")
      .def<static_cast<int (Derived::*)() const>(&Derived::ret1)>("ret1")
      .def<static_cast<int (Derived::*)() const>(&Derived::ret2)>("ret2")
#endif
      .def<&Derived::derived_get>("derived_get");

  m.def<Abstract>("Abstract");
  m.def<DerivedAbstract, Abstract>("DerivedAbstract").cons<int>().def<&DerivedAbstract::Id>("id");

  m.def<&require_Base>("requireBase");
  // Automatic upcasting with multiple inheritance is still not supported
  // (it works only with the first class which is the JS parent)
  m.def<&require_IF1>("requireIF1");
  m.def<&return_Base>("returnBase");
}
